src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.cs

```diff
@@ -1708,326 +1708,326 @@ private bool PropertiesTryRemove(string key)
             switch (key.Length)
             {
                 case 2:
-                    if (_TE.Count > 0
+                    if ((_TE.Count > 0 || !string.IsNullOrEmpty(TE))
                         && string.Equals(key, HeaderNames.TE, StringComparison.Ordinal))
                     {
```

Comment:
This new OR ends up loading the TE header even if that's not what we're trying to remove.

Remove was probably trying to be too smart before, we can change it to act like the others.
```suggestion
                    if (string.Equals(key, HeaderNames.TE, StringComparison.Ordinal))
                    {
                        if (TE.Count > 0)
                        {
                            TE = StringValues.Empty;
                            return true;
                        }
                        return false;
                    }
```
This still has the downside of allocating a string just to check if it's there. A more efficient option would be:
```suggestion
                    if (string.Equals(key, HeaderNames.TE, StringComparison.Ordinal))
                    {
                        if (HasKnownHeader(HttpSysRequestHeader.TE)
                        {
                            TE = StringValues.Empty;
                            return true;
                        }
                        return false;
                    }
```
Where HasKnownHeader would check `bool wasSet = ((_flag1 & 0x2u) != 0);` || if the header exists in the native structure, like GetKnownHeader.
https://github.com/dotnet/aspnetcore/blob/1d031c732419f9999cdb4e9539dc9b020a5957c8/src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs#L313

Reply:
@Tratcher  Are we thinking like that :

```cs
if (string.Equals(key, HeaderNames.TE, StringComparison.Ordinal))
                    {

                        if( ((_flag0 & 0x80u) != 0) || (GetKnownHeader(HttpSysRequestHeader.TE) != null))
                        {
                            TE = StringValues.Empty;
                            return true;
                        }
                        return false;
                    }

```
Please note the following points with this approach:

1. wasSet will be true in all the calls except the first one  if GetKnownHeader returns  Not Null value.
2. GetKnownHeader return value  if it is exist in the native structure and has  initial value (It always return initial value)

So above code will always return true once we sets its value.

```cs
if (string.Equals(key, HeaderNames.TE, StringComparison.Ordinal))
 {
                        if (TE.Count > 0)
                        {
                        }
}

```
This code works fine with all previous scenarios. I will test with new scenarios also.

Note : 

- Previously it was with Underscore now it is without Underscore  (Previous : _TE.Count) (  Now:  TE.Count )
-  _Connection.Count (with underscore) gives 0 even if Connection has initial value.

Please share your views

Reply:
@Tratcher Would you please help me in writing the HasKnownHeader method. I have added the method in RequestHeaders.cs.  Also I made the changes in .tt file (which automatically reflected in the generated file).Please see my Latest Commit.

![image](https://user-images.githubusercontent.com/83835739/202184448-576d00bf-1e97-4e39-9c31-455522e85b87.png)

Reply:
> Would you please help me in writing the HasKnownHeader method.

Follow the implementation of GetKnownHeader down a few layers, you'll likely need to add parallel APIs all the way down. HasKnownHeader should be mostly the same except this line would return true instead of allocating.
https://github.com/dotnet/aspnetcore/blob/442ca2f02e5bfa9b8e5b09de6d088de5ed9e1dc5/src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs#L340

Reply:
@Tratcher  I have done the changes as Suggested and commit my changes. 

The issue with the new code changes is it always return true if there is initial/Native value (Even though we have made the value empty). It always return false if there is no initial/Native value. Please see my comments below and correct me if I am wrong


```cs
headers["Connection"] = "foo";
            bool removedAfterAssignValueIfNativeValue = headers.Remove("Connection");
            bool removedAgainIfNativeValue = headers.Remove("Connection"); // return true. it should be false


            headers["Allow"] = "foo";
            bool removedAfterAssignIfNoNativeValue = headers.Remove("Allow");// return false. it should be true
```